Enh/various analysis fixes and improvements

stephlazar opened the pull request:
Various improvements and bug fixes to analysis_v2 and helper_functions.py.

Changes proposed in this pull request:
- Adds extraction of xlabel and xunit from SweepPoints in RamseyAnalysis and QScaleAnalysis.
- Improves MultiQubit_TimeDomain_Analysis.extract_data() to first check for meas_obj_value_names_map in
   the metadata for extracting channel_map.
- Fixes bug in MultiQubit_TimeDomain_Analysis.get_cal_data_points() and cal_states_analysis() that prevented
   using the analysis without calibration states (i.e. with PCA).
- Adds get_data_from_hdf_file to helper_functions.py.
- Fixes bug in get_data_file_from_timestamp that might have occurred in a merge.

@chellings maybe can review?

Commit:
Improvements and bugfixes in analyses v2 and v3.

Adds extraction of xlabel and xunit from SweepPoints in RamseyAnalysis and QScaleAnalysis.
Improves MultiQubit_TimeDomain_Analysis.extract_data() to first check for meas_obj_value_names_map in
   the metadata for extracting channel_map.
Fixes bug in MultiQubit_TimeDomain_Analysis.get_cal_data_points() and cal_states_analysis() that prevented
   using the analysis without calibration states (i.e. with PCA).
Adds get_data_from_hdf_file to helper_functions.py.
Fixes bug in get_data_file_from_timestamp that might have occurred in a merge.

Commit:
Merge branch 'qudev_master' into enh/various_analysis_fixes_and_improvements

chellings commented on pycqed/analysis_v3/helper_functions.py:
```diff
@@ -68,20 +68,30 @@ def get_param_from_metadata_group(param_name, timestamp=None, data_file=None,
+    return dataset
+
+
 def get_data_file_from_timestamp(timestamp):
```
Since this function returns a handle to an unclosed file, it might make sense to call it open_data_file_from_timestamp() in order to make clear for the user that they open a file and are responsible for closing it properly after use. (As far as I can see, this function is not called from other functions in pycqed and is intended to be used directly from notebooks.)

stephlazar replied:
Ok will rename

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -218,19 +218,22 @@ def __init__(self,
+        self.channel_map = self.get_param_value('meas_obj_value_names_map',
+                                                default_value=dict())
+        if len(self.channel_map) == 0:
+            self.channel_map = self.get_param_value('channel_map')
```
- Shouldn't a channel_map that is manually provided as parameter channel_map have priority over one that is automatically extracted from a different parameter meas_obj_value_names_map?
- Why is it possible to simply assign the meas_obj_value_names_map to self.channel_map? Is a meas_obj_value_names_map and a channel map exactly the same thing? If yes, why do both concepts exist in parallel? If this is related to backwards-compatibility, a comment should clarify what you do and why you are doing it. (Or maybe rather "forward-compatibility"? Do you want to allow applying tda to measurements whose metadata already follows the new paradigms of anav_3?)

stephlazar replied:
Implemented your suggestions

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -457,7 +460,12 @@ def process_data(self):
     def get_cal_data_points(self):
         if len(self.cal_states_rotations) == 0:
-            self.cal_states_dict = {}
+            # this is the case for PCA analysis (no cal points). To use:
```
Instructions to the user ("To use:") probably rather belong into the (non-existent) docstring of the class. If you keep it here for now, I would suggest to make it clearer that this is an instruction for the measurement function. --> "To use PCA, the measurement function has to specify the following in the metadata:"

stephlazar replied:
Added

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -4385,12 +4393,23 @@ def prepare_plots(self):
-                except KeyError:
-                    xlabel = self.raw_data_dict['sweep_parameter_names'][0]
-                    xunit = self.raw_data_dict['sweep_parameter_units'][0]
+                # get x info (try for old sequences which do not have info in meta
```
It seems that this code is replicated from prepare_projected_data_plot in MultiQubit_TimeDomain_Analysis, and it seems that prepare_projected_data_plot is called from inside prepare_plots. Why is it necessary to replicate the code here? (Is this here for the Ramsey plot while the other one if only for the projected plot?) If you need to do this here again, you should consider moving this portion of code in the super class to a separate function, which you can then call from MultiQubit_TimeDomain_Analysis.prepare_projected_data_plot, RamseyAnalysis.prepare_plots, and QScaleAnalysis.prepare_plots.

stephlazar replied:
I will add a get_xaxis_label_unit wrapper.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -501,7 +509,7 @@ def cal_states_analysis(self):
         for qbn in self.qb_names:
             cal_states_dict = self.cal_states_dict_for_rotation[qbn]
-            if len(cal_states_dict) not in [2, 3]:
+            if len(cal_states_dict) > 3:
```
Is this change to allow len 0? Shouldn't it then be
```
if len(cal_states_dict) not in [0, 2, 3]:
```
and the error message should be adapted accordingly?

stephlazar replied:
Good point, I used your suggestion.

chellings reviewed (changes requested):
Thanks for the bugfixes. I just have some questions and comments.

Commit:
Addresses the required changes in pull request #107.

Wraps extraction of xlabel and xunit in a method called get_xaxis_label_unit in MultiQubit_TimeDomain_Analysis and
   uses it in prepare_raw_data_plots, prepare_projected_data_plot, QScaleAnalysis.
Defaults meas_obj_value_names_map to None if not found.
Clarifies comment in get_cal_data_points for the case when len(self.cal_states_rotations) == 0.
Raises error if len(cal_states_dict) not in [0, 2, 3] instead of if len(cal_states_dict) > 3 in cal_states_analysis.

Renames get_data_file_from_timestamp to open_data_file_from_timestamp.
Improves get_param_from_metadata_group to return entire metadata dict if param_name is None.
Adds docstrings to open_data_file_from_timestamp, get_param_from_metadata_group, get_value_names_from_timestamp, get_data_from_hdf_file.

Commit:
Fixes bug that caused QScaleAnalysis to fail when executed twice in a row.

Removes usage of lmfit builtin model functions in QScaleAnalysis.prepare_fitting(), but instead
   creates lambda functions that get converted to lmfit models.

Commit:
Merge branch 'qudev_master' into enh/various_analysis_fixes_and_improvements

chellings committed:
Merge branch 'qudev_master' into enh/various_analysis_fixes_and_improvements

chellings commented on pycqed/analysis_v3/helper_functions.py:
```diff
@@ -68,20 +88,45 @@ def get_param_from_metadata_group(param_name, timestamp=None, data_file=None,
+    else:
+        raise KeyError(f'{Data} was not found in Experimental Data.')
+    if close_file:
+        data_file.close()
```
The file should also be closed if an error occurs (either the KeyError you raise right above or any other error). Same in get_param_from_metadata_group.

stephlazar replied:
Yes :p

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -759,6 +769,29 @@ def rotate_data_TwoD(qb_name, meas_results_per_qb, channel_map,
         return rotated_data_dict
 
+    def get_xaxis_label_unit(self, qb_name):
+        # get xunit and label: temporarily with try catch for old
```
I do not see a try catch ;-)

stephlazar replied:
Outdated comment. I will remove it

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -759,6 +769,29 @@ def rotate_data_TwoD(qb_name, meas_results_per_qb, channel_map,
+            xlabel = sweep_name
+            xunit = sweep_unit
+        else:
+            xlabel = self.raw_data_dict['sweep_parameter_names']
```
Shouldn't we rather use a get with a default value in this else branch to always be on the safe side?

stephlazar replied:
No because these entries will always exist because they are created by the MC and because the extraction of these parameters is hardcoded in the BaseDataAnalysis .

chellings committed:
QScaleAnalysis.prepare_fitting: code formatting and added a comment

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -759,6 +769,29 @@ def rotate_data_TwoD(qb_name, meas_results_per_qb, channel_map,
+            xlabel = self.raw_data_dict['sweep_parameter_names']
+            xunit = self.raw_data_dict['sweep_parameter_units']
+        if np.ndim(xunit) > 0:
+            xunit = xunit[0]
```
In the old version in RamseyAnalysis, it was:
```
                    xlabel = self.raw_data_dict['sweep_parameter_names'][0]
                    xunit = self.raw_data_dict['sweep_parameter_units'][0]
```
Is this also a case that needs to be treated in another branch of if-else? Or maybe by checking the dimension of xlabel as well? (Or by checking whether it is a list instead of a string?)

stephlazar replied:
To be sure, I added the same check for xlabel as for xunit.

Commit:
Merge branch 'qudev_master' into enh/various_analysis_fixes_and_improvements

Commit:
Merge branch 'enh/various_analysis_fixes_and_improvements' of https://github.com/QudevETH/PycQED_py3 into enh/various_analysis_fixes_and_improvements

Commit:
Closes data files in helper_functions.py and improves minor things in get_xaxis_label_unit.
